lib/client/keystore.go

```diff
@@ -687,3 +742,74 @@ func (noLocalKeyStore) SaveCerts(proxy string, cas []auth.TrustedCerts) error {
 }
 func (noLocalKeyStore) GetCerts(proxy string) (*x509.CertPool, error) { return nil, errNoLocalKeyStore }
 func (noLocalKeyStore) GetCertsPEM(proxy string) ([][]byte, error)    { return nil, errNoLocalKeyStore }
+
+// MemLocalKeyStore is an in-memory session keystore implementation.
+type MemLocalKeyStore struct {
+	fsLocalNonSessionKeyStore
+	inMem map[keyIndex]*Key
+}
+
+// NewMemLocalKeyStore initializes a MemLocalKeyStore, the key directory here is only used
+// for storing CA certificates and known host fingerprints.
+func NewMemLocalKeyStore(dirPath string) (*MemLocalKeyStore, error) {
+	dirPath, err := initKeysDir(dirPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	inMem := make(map[keyIndex]*Key)
+	return &MemLocalKeyStore{fsLocalNonSessionKeyStore: fsLocalNonSessionKeyStore{
+		log: logrus.WithFields(logrus.Fields{
+			trace.Component: teleport.ComponentKeyStore,
+		}),
+		KeyDir: dirPath,
+	}, inMem: inMem}, nil
+}
+
+// AddKey writes a key to the underlying key store.
+func (s *MemLocalKeyStore) AddKey(proxy string, username string, key *Key) error {
+	s.inMem[keyIndex{proxyHost: proxy, username: username}] = key
+	if key.ClusterName != "" {
+		s.inMem[keyIndex{proxyHost: proxy, username: username, clusterName: key.ClusterName}] = key
+	}
+	return nil
+}
+
+// GetKey returns the session key for the given username and proxy.
+func (s *MemLocalKeyStore) GetKey(proxy, username string, opts ...KeyOption) (*Key, error) {
+	idx := keyIndex{proxyHost: proxy, username: username}
+	entry, ok := s.inMem[idx]
+	if !ok {
+		return nil, trace.NotFound("key for %v not found", idx)
+	}
+	for _, o := range opts {
+		if err := o.getKey(s, idx, entry); err != nil {
+			s.log.Error(err)
+			return nil, trace.Wrap(err)
+		}
+	}
+	return entry, nil
+}
+
+// DeleteKey removes a specific session key from a proxy.
+func (s *MemLocalKeyStore) DeleteKey(proxyHost, username string, opts ...KeyOption) error {
+	delete(s.inMem, keyIndex{proxyHost: proxyHost, username: username})
```

Comment:
should this also delete any keys where `clusterName` is set?

Reply:
Once the `opts` handling is added here (see my comment below) leaving some `clusterName` leftovers shouldn't break anything as the only way to get a key with a `clusterName` is to go via an `opt`.

Alternatively the map's key could be just changed from `[proxyHost, username, clusterName]` to `[proxyHost, username][clusterName]`.

Reply:
This should be resolved now with the opts handling.